muikku-core-plugins/src/main/java/fi/otavanopisto/muikku/plugins/communicator/rest/CommunicatorRecipientsRESTService.java

```diff
@@ -0,0 +1,308 @@
+    ) {
+    
+    if (!sessionController.isLoggedIn()) {
+      return Response.status(Status.FORBIDDEN).build();
+    }
+
+    UserEntity loggedUser = sessionController.getLoggedUserEntity();
+    
+    EnvironmentUser environmentUser = environmentUserController.findEnvironmentUserByUserEntity(loggedUser);
+    EnvironmentRoleArchetype loggedUserRole = (environmentUser != null && environmentUser.getRole() != null) ? 
+        environmentUser.getRole().getArchetype() : null;
+    
+    Set<Long> userGroupFilters = null;
+    Set<Long> workspaceFilters = null;
+    Set<EnvironmentRoleArchetype> roleArchetypeFilter = new HashSet<>();
+    
+    if (EnvironmentRoleArchetype.STUDENT.equals(loggedUserRole)) {
+      // Stuff students can seach for
+      roleArchetypeFilter.add(EnvironmentRoleArchetype.ADMINISTRATOR);
+      roleArchetypeFilter.add(EnvironmentRoleArchetype.MANAGER);
+      roleArchetypeFilter.add(EnvironmentRoleArchetype.STUDY_PROGRAMME_LEADER);
+      roleArchetypeFilter.add(EnvironmentRoleArchetype.TEACHER);
+      roleArchetypeFilter.add(EnvironmentRoleArchetype.STUDY_GUIDER);
+    } else {
+      // Default for other roles
+      
+      roleArchetypeFilter.add(EnvironmentRoleArchetype.ADMINISTRATOR);
+      roleArchetypeFilter.add(EnvironmentRoleArchetype.MANAGER);
+      roleArchetypeFilter.add(EnvironmentRoleArchetype.STUDY_PROGRAMME_LEADER);
+      roleArchetypeFilter.add(EnvironmentRoleArchetype.TEACHER);
+      roleArchetypeFilter.add(EnvironmentRoleArchetype.STUDY_GUIDER);
+      roleArchetypeFilter.add(EnvironmentRoleArchetype.STUDENT);
+      
+      if (sessionController.hasEnvironmentPermission(RoleFeatures.ACCESS_ONLY_GROUP_STUDENTS)) {
+        // Study guider limitations - groups where user is a member
+        List<UserGroupEntity> userGroups = userGroupEntityController.listUserGroupsByUserIdentifier(sessionController.getLoggedUser());
+        userGroupFilters = userGroups.stream().map(userGroup -> userGroup.getId()).collect(Collectors.toSet());
+      }
+    }
+
+    SearchProvider searchProvider = getSearchProvider();
+    if (searchProvider != null) {
+      String[] fields = new String[] { "firstName", "lastName", "nickName", "email" };
+
+      Boolean onlyDefaultUsers = true;
+      
+      SearchResult result = searchProvider.searchUsers(searchString, 
+          fields, 
+          roleArchetypeFilter, 
+          userGroupFilters, 
+          workspaceFilters, 
+          null, 
+          false,
+          false,
+          onlyDefaultUsers,
+          firstResult, 
+          maxResults);
+      
+      List<Map<String, Object>> results = result.getResults();
+
+      List<fi.otavanopisto.muikku.rest.model.User> ret = new ArrayList<fi.otavanopisto.muikku.rest.model.User>();
+
+      for (Map<String, Object> o : results) {
+        String[] id = ((String) o.get("id")).split("/", 2);
+        UserEntity userEntity = userEntityController
```

Comment:
Users in the search index also have userEntityId so could be slightly simpler to find with that?